### DISCUSSION TITLE
Authentication: login not working

### DISCUSSION THREAD

**Discussion by gtamas** [loopbackio/loopback-next#7232] [Q&A] [answered]
I added JWT auth to my app, tried to follow your docs.

I have a custom UserRepository and UserService class. 

I have this is application.ts:

```
   this.dataSource(MongoDataSource, UserServiceBindings.DATASOURCE_NAME);
    this.bind(UserServiceBindings.USER_SERVICE).toClass(MyUserService);
    this.bind(UserServiceBindings.USER_REPOSITORY).toClass(MyUserRepository);

    this.component(AuthenticationComponent);
```

Yet, when calling login I'm getting this error:

```
Request POST /users/login failed with status code 500. TypeError: **this.userRepository.findCredentials** is not a function
    at MyUserService.verifyCredentials (/path/to/project/user-service/node_modules/@loopback/authentication-jwt/src/services/user.service.ts:38:56)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

So even though I've bound my own implementation of UserService, it's still trying to use the default one for some reason. What's wrong here?

MyUserService points to ./services/user-service.ts but it's not used for some reason. Any help please? 

I wanna use my own UserService and UserRepository.

THanks

**Comment by gtamas** [comment, marked as answer]
The solution was to move the binding calls to the bottom of the constructor, after JWTAuthenticationComponent registration.
I don't get this.

There is only one call to:

`  this.bind(UserServiceBindings.USER_SERVICE)`

Does it matter when do I call this function? I thought this is not important. Apparently, this has to be called after JWTAuthenticationComponent bindings are registered. 

I don't remember reading anything about this in the docs. What am I missing here?

**Reply by achrinza** [reply]
Components have several well-defined public properties. Most notably, `bindings`. `bindings` is an array of Bindings that will be bound to the Application Context when registering the Component using `this.component()`; This can be thought of as executing `this.bind()` at the moment `this.component()` is called.

When registering `JWTAuthenticationComponent`, the default bindings are registered-Including those for `UserServiceBindings.USER_SERVICE` and `UserServiceBindings.USER_REPOSITORY`.

https://github.com/strongloop/loopback-next/blob/5d0594846deeb8c2e82fd5312465692833fe4176/extensions/authentication-jwt/src/jwt-authentication-component.ts#L44-L45

Hence, calling `this.component()` will cause conflicting Bindings to be overwritten.

**Reply by gtamas** [reply]
Ah Ok! I see, thanks for the explanation!
